go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -165,59 +178,138 @@ func (oc *Controller) updateEgressFirewallWithRetry(egressfirewall *egressfirewa
+			}
+			oc.egressFirewallDNS.Add(ef.namespace, rule.to.dnsName, dnsNameAddressSets)
+			match, err = generateMatch(hashedAddressSetNameIPv4, hashedAddressSetNameIPv6, []string{dnsNameAddressSets.GetIPv4HashName(), dnsNameAddressSets.GetIPv6HashName()}, rule.ports)
+			if err != nil {
+				return err
+			}
 		}
-
-		if len(rule.ports) > 0 {
-			match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], egressGetL4Match(rule.ports))
+		for _, joinSwitch := range joinSwitches {
+			err := createACLRule(defaultStartPriority-rule.id, match, action, ef.namespace, joinSwitch)
+			if err != nil {
+				return err
+			}
 		}
-		uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
-			"--columns=_uuid", "find", "ACL", match, "action="+action,
-			fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace))
+	}
+	return nil
+}
+
+func createACLRule(priority int, match, action, namespace, joinSwitch string) error {
+	uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=_uuid", "find", "ACL", match, "action="+action,
+		fmt.Sprintf("external-ids:egressFirewall=%s", namespace))
 
+	if err != nil {
+		return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+	}
+	if uuid == "" {
+		_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
+			fmt.Sprintf("priority=%d", priority),
+			fmt.Sprintf("direction=%s", fromLport),
+			match, "action="+action,
+			fmt.Sprintf("external-ids:egressFirewall=%s", namespace),
+			"--", "add", "logical_switch", joinSwitch,
+			"acls", "@acl")
 		if err != nil {
-			return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+			return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
 		}
-		for _, joinSwitch := range joinSwitches {
-			if uuid == "" {
-				_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
-					fmt.Sprintf("priority=%d", defaultStartPriority-rule.id),
-					fmt.Sprintf("direction=%s", fromLport), match, "action="+action,
-					fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace),
-					"--", "add", "logical_switch", joinSwitch,
-					"acls", "@acl")
-				if err != nil {
-					return fmt.Errorf("error executing create ACL command, stderr: %q, %+v", stderr, err)
-				}
-			} else {
-				_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
-				if err != nil {
-					return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
+	} else {
+		_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
+		if err != nil {
+			return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
 
-				}
-			}
 		}
 	}
 	return nil
 }
 
+func generateMatch(ipv4Source, ipv6Source string, destinations []string, ports []egressfirewallapi.EgressFirewallPort) (string, error) {
```

Comment:
please add some kind of docblock; this is a complicated function.

Reply:
added some comments PTAL